Diego Lozoya Morales | 745345

01/09/23

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

1. Criterio de optimización
- minimizar
2. Variables sobre las cuales tengo control
+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel
3. Función de costos
$$ 0.013x_1 +  0.008x_2 + 0.010x_3 + 0.002x_4 + 0.005x_5 + 0.001x_6 = costo $$
4. Restricciones 
- Cantidad = 100
    - $ 1x_1 + 1x_2 + 1x_3 + 1x_4 + 1x_5 + 1x_6 = 100 $
- Proteína $\geq$ 8
    - $ 0.1x_1 + 0.2x_2 + 0.15x_3 + 0x_4 + 0.04x_5 + 0x_6 \geq 8 $
- Grasa $\geq$ 6
    - $ 0.08x_1 + 0.1x_2 + 0.11x_3 + 0.01x_4 + 0.01x_5 + 0x_6 \geq 6 $
- Fibra $\leq$ 2
    - $ 0.001x_1 + 0.005x_2 + 0.005x_3 + 0.1x_4 + 0.15x_5 + 0x_6 \leq 2 $
- Sal $\leq$ 0.4
    - $ 0.002x_1 + 0.005x_2 + 0.007x_3 + 0.002x_4 + 0.008x_5 + 0x_6 \leq 0.4 $
5. Solución

In [2]:
import scipy.optimize as opt
import numpy as np

c = np.array([0.013,
              0.008,
              0.010,
              0.002,
              0.005,
              0.001])

Aub = np.array([
    [-0.10, -0.20, -0.15, -0.00, -0.04, -0.00],
    [-0.08, -0.10, -0.11, -0.01, -0.01, -0.00],
    [0.001, 0.005, 0.005, 0.100, 0.150, 0.000],
    [0.002, 0.005, 0.007, 0.002, 0.008, 0.000]
])

bub = np.array([-8,
                -6,
                2.0,
                0.4])

Aeq = np.array([
    [1, 1, 1, 1, 1, 1]
])

beq = np.array([100])

res = opt.linprog(c, Aub, bub, Aeq, beq)
res.x

array([ 0., 60.,  0.,  0.,  0., 40.])

In [3]:
res.fun

0.52

Para cumplir con las medidas nutimentales se debe de hacer una mezcla de 60g de carne y 40g de gel por cada 100g de alimento, para minimizar los gastos, gastando un total de $\$0.52$ por 100g

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

1. Criterio de optimización
- minimizar
2. Variables sobre las cuales tengo control
+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia
3. Función de costos
$$ 5x_1 +  2x_2 + 7x_3 + 3x_4 + 6x_5 + 6x_6 + 6x_7 + 1x_8 + 2x_9 = costo $$
4. Restricciones 
- Producción planta 1 $\leq$ 80
    - $ 1x_1 + 1x_2 + 1x_3 + 0x_4 + 0x_5 + 0x_6 + 0x_7 + 0x_8 + 0x_9 \leq 80 $
- Producción planta 2 $\leq$ 40
    - $ 0x_1 + 0x_2 + 0x_3 + 1x_4 + 1x_5 + 1x_6 + 0x_7 + 0x_8 + 0x_9 \leq 40 $
- Producción planta 3 $\leq$ 60
    - $ 0x_1 + 0x_2 + 0x_3 + 0x_4 + 0x_5 + 0x_6 + 1x_7 + 1x_8 + 1x_9 \leq 60 $
- Demanda Guadalajara = 70
    - $ 1x_1 + 0x_2 + 0x_3 + 1x_4 + 0x_5 + 0x_6 + 1x_7 + 0x_8 + 0x_9 = 70 $
- Demanda León = 40
    - $ 0x_1 + 1x_2 + 0x_3 + 0x_4 + 1x_5 + 0x_6 + 0x_7 + 1x_8 + 0x_9 = 40 $
- Demanda Morelia = 70
    - $ 0x_1 + 0x_2 + 1x_3 + 0x_4 + 0x_5 + 1x_6 + 0x_7 + 0x_8 + 1x_9 = 70 $
5. Solución

In [4]:
c = np.array([5,
              2,
              7,
              3,
              6,
              6,
              6,
              1,
              2])

Aub = np.array([
    [1, 1, 1, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 1, 1, 1, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 1, 1, 1]
])

bub = np.array([80,
                40,
                60])

Aeq = np.array([
    [1, 0, 0, 1, 0, 0, 1, 0, 0],
    [0, 1, 0, 0, 1, 0, 0, 1, 0],
    [0, 0, 1, 0, 0, 1, 0, 0, 1]
])

beq = np.array([70,
                40,
                70])

res = opt.linprog(c, Aub, bub, Aeq, beq)
res.x

array([30., 40., 10., 40.,  0.,  0.,  0.,  0., 60.])

In [5]:
res.fun

540.0

Para cumplir con la demanda de las ciudades, dependiendo de la producción por planta, se deben de mandar 30 millones de Kw de la planta 1 a Guadalajara, 40 millones de Kw de la planta 1 a León, 10 millones de Kw de la planta 1 a Morelia, 40 millones de Kw de la planta 2 a Guadalajara y 60 millones de Kw de la planta 3 a Morelia, para gastar lo mínimo, lo cual es 540 unidades monetarias.